clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -2833,7 +2833,35 @@ static void handleWeakImportAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
   D->addAttr(::new (S.Context) WeakImportAttr(S.Context, AL));
 }
 
-// Handles reqd_work_group_size and work_group_size_hint.
+// Checks correctness of mutual usage of different work_group_size attributes:
+// reqd_work_group_size, max_work_group_size
+static bool checkWorkGroupSizeValues(Sema &S, Decl *D, const ParsedAttr &Attr,
```

Comment:
The caller being `template <typename WorkGroupAttr>`, I would strongly prefer to see this function template as well.

Reply:
A close future plans:
I will add one more kernel attribute that will interact with both reqd_wg_size and max_wg_size. The function will check the correctness of attributes' arguments depending on each other. For example, currently max_wg_size limits possible values for reqd_wg_size arguments and this function is checking it. With a new attribute added, if it accepts '0' in its single argument, max_wg_size should have (1,1,1) passed which limits reqd_wg_size parameters by ([0,1] , [0,1] , [0,1]). All the handling will be implemented in this function and it will be used only for them. So the function will expand and at the same time no code duplication is expected.

To conclude:
I don't see a reason to make the function to be template function.

Reply:
Thanks, fine with me then.